### PR TITLE
GC Fixes for Upstream Update

### PR DESCRIPTION
**Pull request by bholmes**
Need to implement a simple version of Frozen Segments

Adding "FEATURE_UNITY_NULLGC_WRITEBARRIER_PATCH"
to workaround write barrier problems with our GC. We need to
remove FEATURE_UNITY_NULLGC_WRITEBARRIER_PATCH
completely  when we move back to the default GC.

### CODE REVIEW

File: src/coreclr/vm/amd64/jithelpers_fastwritebarriers.S [hunk 1]
```diff
@@ -32,6 +32,23 @@ PATCH_LABEL JIT_WriteBarrier_PreGrow64_Patch_Label_Lower
...
 
+#ifdef FEATURE_UNITY_NULLGC_WRITEBARRIER_PATCH
+
+PATCH_LABEL JIT_WriteBarrier_PreGrow64_Patch_Label_Upper
+        movabs  r8, 0xF0F0F0F0F0F0F0F0
+
+        cmp     rsi, r8
+
+#ifdef FEATURE_MANUALLY_MANAGED_CARD_BUNDLES
+        .byte 0x73, 0x4b
```
Comment (joncham) [inline src/coreclr/vm/amd64/jithelpers_fastwritebarriers.S]: Is this a jump instruction?

**Review by joncham — APPROVED** [review]

File: src/coreclr/vm/amd64/jithelpers_fastwritebarriers.S [hunk 2]
```diff
@@ -415,6 +432,24 @@ PATCH_LABEL JIT_WriteBarrier_WriteWatch_PreGrow64_Patch_Label_Lower
...
+
+#ifdef FEATURE_UNITY_NULLGC_WRITEBARRIER_PATCH
+
+PATCH_LABEL JIT_WriteBarrier_WriteWatch_PreGrow64_Patch_Label_Upper
+        movabs  r10, 0xF0F0F0F0F0F0F0F0
+
+        cmp     rsi, r10
+
+#ifdef FEATURE_MANUALLY_MANAGED_CARD_BUNDLES
+        .byte 0x73, 0x3b
```
Comment (bholmes) [inline src/coreclr/vm/amd64/jithelpers_fastwritebarriers.S]: The offset of 0x3b is probably wrong as there is an extra shr instruction on line 468.

Not sure what I am going to do about it as it may not be used.

Comment (bholmes) [reply src/coreclr/vm/amd64/jithelpers_fastwritebarriers.S]: I think I have a better way to do this...

TLDR;
We need to force the bReqUpperBoundsCheck bool to be true here...

https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/amd64/jitinterfaceamd64.cpp#L833

Will swap this out and try again.

**Comment by bholmes** [comment]
What I think is going on with the tests...
 - Windows x86 is broken upstream
 - OSX amd64 May be related to my Write barrier patch, will try something else
 - OSX arm64 (not run on the server) Failing, probably due to WB code that needs attention for arm
